src/net_processing.cpp

```diff
@@ -94,144 +94,156 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
-                        // chain if they are valid, and no more than a month older (both in time, and in
-                        // best equivalent proof of work) than the best header chain we know about.
-                        fSend = mi->IsValid(BLOCK_VALID_SCRIPTS) && (pindexBestHeader != NULL) &&
-                                (pindexBestHeader.load()->GetBlockTime() - mi->GetBlockTime() < nOneMonth) &&
-                                (GetBlockProofEquivalentTime(
-                                     *pindexBestHeader, *mi, *pindexBestHeader, consensusParams) < nOneMonth);
-                        if (!fSend)
+                        LOCK(cs_main);
+                        if (chainActive.Contains(mi))
                         {
-                            LOG(NET, "%s: ignoring request from peer=%s for old block that isn't in the main chain\n",
-                                __func__, pfrom->GetLogName());
+                            fSend = true;
                         }
                         else
                         {
-                            // BU: don't relay excessive blocks that are not on the active chain
-                            if (mi->nStatus & BLOCK_EXCESSIVE)
-                                fSend = false;
+                            READLOCK(cs_mapBlockIndex);
+                            static const int nOneMonth = 30 * 24 * 60 * 60;
+                            // To prevent fingerprinting attacks, only send blocks outside of the active
+                            // chain if they are valid, and no more than a month older (both in time, and in
+                            // best equivalent proof of work) than the best header chain we know about.
+                            fSend = mi->IsValid(BLOCK_VALID_SCRIPTS) && (pindexBestHeader != NULL) &&
+                                    (pindexBestHeader.load()->GetBlockTime() - mi->GetBlockTime() < nOneMonth) &&
+                                    (GetBlockProofEquivalentTime(
+                                         *pindexBestHeader, *mi, *pindexBestHeader, consensusParams) < nOneMonth);
                             if (!fSend)
-                                LOG(NET, "%s: ignoring request from peer=%s for excessive block of height %d not on "
-                                         "the main chain\n",
-                                    __func__, pfrom->GetLogName(), mi->nHeight);
+                            {
+                                LOG(NET,
+                                    "%s: ignoring request from peer=%s for old block that isn't in the main chain\n",
+                                    __func__, pfrom->GetLogName());
+                            }
+                            else
+                            {
+                                // BU: don't relay excessive blocks that are not on the active chain
+                                if (mi->nStatus & BLOCK_EXCESSIVE)
```

Comment:
the BLOCK_EXCESSIVE bit doesn't change so I don't think we need the lock, and even if it did change, worst case we relay an excessive block...

Reply:
we still need the lock for mi->IsValid() further up, but I can reduce the scope of the lock.